internal/policy/verify.go

```diff
@@ -108,52 +91,66 @@ func VerifyRefFromEntry(ctx context.Context, repo *gitinterface.Repository, targ
 		return gitinterface.ZeroHash, err
 	}
 
-	// Find policy entry before the starting point entry
-	slog.Debug("Identifying applicable policy entry...")
-	policyEntry, _, err := rsl.GetLatestReferenceEntryForRefBefore(repo, PolicyRef, fromEntry.GetID())
-	if err != nil {
-		return gitinterface.ZeroHash, err
-	}
-
-	slog.Debug("Identifying applicable attestations entry...")
-	var attestationsEntry *rsl.ReferenceEntry
-	attestationsEntry, _, err = rsl.GetLatestReferenceEntryForRefBefore(repo, attestations.Ref, fromEntry.GetID())
-	if err != nil {
-		if !errors.Is(err, rsl.ErrRSLEntryNotFound) {
-			return gitinterface.ZeroHash, err
-		}
-	}
-
 	// Do a relative verify from start entry to the latest entry
 	slog.Debug("Verifying all entries...")
-	return latestEntry.TargetID, VerifyRelativeForRef(ctx, repo, policyEntry, attestationsEntry, fromEntry, latestEntry, target)
+	return latestEntry.TargetID, VerifyRelativeForRef(ctx, repo, fromEntry, latestEntry, target)
 }
 
 // VerifyRelativeForRef verifies the RSL between specified start and end entries
 // using the provided policy entry for the first entry.
-//
-// TODO: should the policy entry be inferred from the specified first entry?
-func VerifyRelativeForRef(ctx context.Context, repo *gitinterface.Repository, initialPolicyEntry, initialAttestationsEntry, firstEntry, lastEntry *rsl.ReferenceEntry, target string) error {
+func VerifyRelativeForRef(ctx context.Context, repo *gitinterface.Repository, firstEntry, lastEntry *rsl.ReferenceEntry, target string) error {
 	var (
 		currentPolicy       *State
 		currentAttestations *attestations.Attestations
+		err                 error
 	)
 
 	// Load policy applicable at firstEntry
-	slog.Debug("Loading initial policy...")
-	state, err := LoadState(ctx, repo, initialPolicyEntry)
-	if err != nil {
-		return err
+	// If firstEntry is for policy, we don't walk the RSL here
+	slog.Debug(fmt.Sprintf("Loading policy applicable at first entry '%s'...", firstEntry.ID))
+	var initialPolicyEntry *rsl.ReferenceEntry
+	if firstEntry.RefName == PolicyRef {
+		initialPolicyEntry = firstEntry
+	} else {
+		initialPolicyEntry, _, err = rsl.GetLatestReferenceEntryForRefBefore(repo, PolicyRef, firstEntry.ID)
```

Comment:
This is a more explicit change from what we were doing before because now the start point for verification can be anywhere. Earlier, we had an implicit assumption it'd be from start of history.

So now, if the first entry is for policy, it's loaded to verify subsequent entries. If the first entry is not for policy and there's no policy entry before it, then the first entry must be _the very first entry_ in the RSL. So this supports the earlier case, but now checks explicitly.

Also, another change is how the very first entry case is handled. The odds are that the very first entry is for `policy-staging`, which we previously loaded and switched to the `policy` entry when it was encountered subsequently. Now, this isn't the case. We leave the policy unset so that `currentPolicy` is only ever set to an entry that was "applied" into `refs/gittuf/policy`. If everything is well, currentPolicy is set before any real entry is verified. If it's not set, then we error saying policy wasn't found. This closes a bug we previously had where an unapplied policy could be used to verify some RSL entry if the policy was not applied before said entry was verified.

Reply:
Just realized my TODO to add a test checking for this case wasn't addressed where we have policy-staging <- main.

Reply:
As it's written now, it'd fail. We could loosen this requirement: just leave policy unset so if the workflow encounters a non refs/gittuf ref entry, verification still fails as we didn't see a refs/gittuf/policy first.

Reply:
I've taken out the check for now.